compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt

```diff
@@ -867,150 +805,98 @@ class ExpressionCodegen(
         mv.fakeAlwaysFalseIfeq(endLabel)
+        data.withBlock(LoopInfo(loop, continueLabel, endLabel)) {
+            loop.body?.accept(this, data)?.discard()
+        }
+        mv.visitLabel(continueLabel)
+        loop.condition.markLineNumber(true)
+        loop.condition.accept(this, data).coerceToBoolean().jumpIfTrue(entry)
+        mv.mark(endLabel)
+        return voidValue
+    }
 
-        with(LoopInfo(loop, continueLabel, endLabel)) {
-            data.addInfo(this)
-            loop.body?.let {
-                gen(it, data).discard()
+    private fun unwindBlockStack(endLabel: Label, data: BlockInfo, loop: IrLoop? = null): LoopInfo? {
+        return data.handleBlock {
+            when {
+                it is TryInfo -> {
+                    it.gaps.add(markNewLabel() to endLabel)
+                    it.onExit.accept(this, data).discard()
+                }
+                it is LoopInfo && it.loop == loop -> return it
             }
-            data.removeInfo(this)
+            unwindBlockStack(endLabel, data, loop)
         }
+    }
 
-        mv.visitLabel(continueLabel)
-        generateLoopJump(loop.condition, data, entry, false)
+    override fun visitBreakContinue(jump: IrBreakContinue, data: BlockInfo): PromisedValue {
+        jump.markLineNumber(startOffset = true)
+        val endLabel = Label()
+        val stackElement = unwindBlockStack(endLabel, data, jump.loop)
+            ?: throw AssertionError("Target label for break/continue not found")
+        mv.fixStackAndJump(if (jump is IrBreak) stackElement.breakLabel else stackElement.continueLabel)
         mv.mark(endLabel)
-
-        return StackValue.none()
+        return voidValue
     }
 
-    override fun visitTry(aTry: IrTry, data: BlockInfo): StackValue {
+    override fun visitTry(aTry: IrTry, data: BlockInfo): PromisedValue {
         aTry.markLineNumber(startOffset = true)
-        val finallyExpression = aTry.finallyExpression
-        val tryInfo = if (finallyExpression != null) TryInfo(aTry) else null
-        if (tryInfo != null) {
-            data.addInfo(tryInfo)
-        }
+        return if (aTry.finallyExpression != null)
+            data.withBlock(TryInfo(aTry.finallyExpression!!)) { visitTryWithInfo(aTry, data, it) }
+        else
+            visitTryWithInfo(aTry, data, null)
+    }
 
+    private fun visitTryWithInfo(aTry: IrTry, data: BlockInfo, tryInfo: TryInfo?): PromisedValue {
         val tryBlockStart = markNewLabel()
         mv.nop()
-        gen(aTry.tryResult, aTry.asmType, data)
+        val tryResult = aTry.tryResult.accept(this, data).coerce(aTry.asmType).materialized
         val tryBlockEnd = markNewLabel()
-
-        val tryRegions = getCurrentTryIntervals(tryInfo, tryBlockStart, tryBlockEnd)
-
+        val tryBlockGaps = tryInfo?.gaps?.toList() ?: listOf()
         val tryCatchBlockEnd = Label()
-        genFinallyBlockOrGoto(tryInfo, tryCatchBlockEnd, null, data)
+        mv.goTo(tryCatchBlockEnd)
```

Comment:
@pyos It's best to avoid any changes in finally block scheme generation. Old scheme is necessary for Inliner to  process non-locals returns (they are not supported yet in IR backend, testData/codegen/boxInline/nonLocalReturns).
Similar problem could be with new logic of gap interval processing. 
So I would prefer to posptone these to changes until non-local return support in Inliner.

Reply:
Yes, I've noticed that, which is why there's an additional insertion of an empty gap (i.e. a TRYCATCHBLOCK that covers nothing and jumps to itself) at line 934. This is enough for the inliner to detect the finally block, although it seems to not work anyway because of a missing constant load of what I assume is the length of the finally right after ASTORE.

Reply:
@pyos I'm now in process of fixing basic non-local returns in inliner. It would be new fails with changed logic around finally generations and gaps